Use SourceManager in engine

mcastorina opened the pull request:
<!--
Please create an issue to collect feedback prior to feature additions. Please also reference that issue in any PRs.
If possible try to keep PRs scoped to one feature, and add tests for new features.
-->

mcastorina committed:
Add SourceManager to Engine struct

mcastorina committed:
Update Engine methods to use the SourceManager

mcastorina commented:
https://github.com/trufflesecurity/trufflehog/pull/1518 should be reviewed / merged first. TIA

rosecodym commented:
if you set the base branch of this pr to be your other pr then this pr's diff won't include that pr's changes. is that what you're going for here?

mcastorina committed:
Fix GCS test

The original was testing that `Init()` errors weren't surfaced in
`Finish()`, but the `SourceManager` changed that behavior.

mcastorina committed:
JobProgress race fixes

mcastorina committed:
Add contextual values

mcastorina committed:
Remove unused code

mcastorina committed:
Add debug logs

rosecodym commented on pkg/sources/source_manager.go:
```diff
@@ -232,7 +249,10 @@ func (s *SourceManager) runWithoutUnits(ctx context.Context, handle handle, sour
 			report.ReportChunk(nil, chunk)
-			_ = common.CancellableWrite(ctx, s.outputChunks, chunk)
+			if err := common.CancellableWrite(ctx, s.outputChunks, chunk); err != nil {
+				report.ReportError(Fatal{err})
```
What makes this fatal?

mcastorina replied:
I have somewhat arbitrarily chosen to make context cancellations fatal errors so that they will surface as an error from a run, unlike errors we normally encounter but continue scanning.

ahrav replied:
question: What happens if common.CancellableWrite encounters an error?

Looks like the error is reported by calling report.ReportError(Fatal{err}), and then the goroutine is immediately exited with return.

A potential problem here is that runWithoutUnits is not directly aware of this change. The error is reported via report.ReportError, but unless ReportError has side effects that the main function checks (which i hope isn't the case), I think this fxn will continue running.

In which case the source.Chunks function might continue to send chunks to the channel, but there is no longer a consumer for the channel since the goroutine has exited. I think?

ahrav replied:
Maybe use an error channel to communicate errors from the goroutine back to the main function?

rosecodym commented on pkg/sources/job_progress.go:
```diff
@@ -163,11 +166,15 @@ func (jp *JobProgress) Finish() {
-	atomic.AddUint64(&jp.metrics.TotalUnits, 1)
+	jp.metricsLock.Lock()
+	jp.metrics.TotalUnits++
+	jp.metricsLock.Unlock()
```
Why aren't atomics enough here?

rosecodym replied:
ah i see, there's a read elsewhere that this needs to guard against

rosecodym reviewed (approved):
some non-blocking questions but this looks like a pretty smooth transition

mcastorina commented on pkg/engine/engine.go:
```diff
@@ -276,8 +272,11 @@ func Start(ctx context.Context, options ...EngineOption) (*Engine, error) {
-	e.sourcesWg.SetLimit(int(e.concurrency))
+	// Create SourceManager.
+	e.sourceManager = sources.NewManager(
+		sources.WithConcurrency(int(e.concurrency)),
+		sources.WithConcurrentUnits(int(e.concurrency)),
```
Great point, I'll update it to `WithConcurrentSources`

mcastorina commented on pkg/sources/job_progress.go:
```diff
@@ -196,7 +205,12 @@ func (jp *JobProgress) EndEnumerating(end time.Time) {
+	metrics.Errors = make([]error, len(metrics.Errors))
+	copy(metrics.Errors, jp.metrics.Errors)
+
+	return metrics
```
Not that I'm aware of 😕

mcastorina commented on pkg/engine/gcs_test.go:
```diff
@@ -80,7 +80,10 @@ func TestScanGCS(t *testing.T) {
 				return
 			}
-			assert.Nil(t, e.Finish(ctx))
+			if err := e.Finish(ctx); err != nil && !test.wantErr && !strings.Contains(err.Error(), "googleapi: Error 400: Bad Request") {
+				t.Errorf("Finish() got: %v, want: %v", err, nil)
+				return
+			}
```
```
The original was testing that `Init()` errors weren't surfaced in
`Finish()`, but the `SourceManager` changed that behavior.
```

(also included in the commit message for these changes)

mcastorina committed:
Rename WithConcurrency to WithConcurrentSources

mcastorina committed:
Always forward chunks to the output chunks channel
